### DISCUSSION TITLE
Creating a single-line field (disabling Enter)

### DISCUSSION THREAD

**Discussion by jimirobaer** [craftcms/ckeditor#99] [Ideas]
We are in the process of converting existing fields from Redactor to CKEditor.

One of our fields disabled the Enter Key (using 'enterKey: false') which made the field act essentially as 'single line' field. We use this type of field quite often for text fields that span multiple lines, but not paragraphs.

It seems impossible to recreate the Redactor behaviour 100%, which would be the following:

1. Remove any paragraphs and convert everything to line breaks.
2. Disable Enter (Shift+Enter is allowed) in the editor.

The first option is solved using an updated HTML Purifier configuration. The second option does not seem possible out-of-the-box. We investigated a few things:

1. CKEditor uses plugins for Enter & ShiftEnter functionality, but it's not possible to disable them using `removePlugins` ([API reference](https://ckeditor.com/docs/ckeditor5/latest/api/module_core_editor_editorconfig-EditorConfig.html#member-removePlugins)).
2. It's not possible to capture and disable enterKey events using the config options currently available from the CMS.  ([workaround?)](https://github.com/ckeditor/ckeditor5/issues/9422#issuecomment-814727348)

Anyone created such a field using CKEditor 5 before? Any clue how to approach this?

**Comment by msimpson** [comment]
I'm looking for similar functionality and I believe [Restricted Editing](https://ckeditor.com/docs/ckeditor5/latest/features/restricted-editing.html) is a way to go.

It should be possible to add a [DLL build](https://ckeditor.com/docs/ckeditor5/latest/installation/advanced/alternative-setups/dll-builds.html#:~:text=CKEditor%C2%A05%20comes%20with%20ready%2Dto%2Duse%20DLL%20builds.%20These%20builds%20are%20added%20to%20the%20npm%20packages%20and%20they%20are%20available%20inside%20the%20/build%20directory%20of%20each%20package.) of [@ckeditor/ckeditor5-restricted-editing (`build/restricted-editing.js`)](https://www.npmjs.com/package/@ckeditor/ckeditor5-restricted-editing?activeTab=code) as [described in the README](https://github.com/craftcms/ckeditor?tab=readme-ov-file#adding-ckeditor-plugins) using a Craft [plugin](https://craftcms.com/docs/5.x/extend/plugin-guide.html) or [module](https://craftcms.com/docs/5.x/extend/module-guide.html).

The only trick will be getting a bit of templated content, like, `<span class="restricted-editing-exception">Lorem ipsum dolor.</span>`, added to the field by default (without stomping on existing content).

There is an [`initialData`](https://ckeditor.com/docs/ckeditor5/latest/api/module_core_editor_editorconfig-EditorConfig.html#member-initialData) configuration option that could be used, but I've found it overwrites saved data repopulating the field. So, going this route will require something more.

Perhaps hooking into the fields with JavaScript and checking for existing content before asserting a default value. Or possibly writing a third-party CKEditor plugin with similar logic so a configuration option could be provided, like `initialData`, only if the field is blank upon load.

In any case, this is all conjecture until I sit down with it.

EDIT: [This issue](https://github.com/ckeditor/ckeditor5/issues/762) is about the closest thing to this request, and it's all still up in the air.

For the time being, I'm going to try a restricted toolbar of inline formatting only (bold, italic, underline, etc.) paired with [striptags](https://twig.symfony.com/doc/3.x/filters/striptags.html). Similar to [mlncn's workaround for Drupal (in the above issue)](https://github.com/ckeditor/ckeditor5/issues/762#issuecomment-1908034712). Like: `{{ entry.richText | striptags( '<b><i><em><del><span>' ) }}`.

**Comment by mihob** [comment]
Is there now a way to format a single line with the CKEditor plugin? This was very easy to do with Redactor. @brandonkelly @i-just 

We use this for many customers, for example to further format a single heading, etc.  And we would now like to finally migrate to CKEditor, especially in view of a future Craft CMS 6 release and possible incompatibility with Redactor. However, this often fails because even simple adjustments are not possible with the CKEditor plugin.

What exactly do we want to customise?
- Hide normal toolbar
- Use balloon toolbar (CKEditor provides this, but the extension seems to be missing in the CKEditor plugin)
- Prevent line breaks.

**Comment by msimpson** [comment]
My current implementation is using an HTML purifier configuration to whittle down the allowed tags:

```
{
  "HTML.Allowed":  "em,i" 
}
```

And using a minimal toolbar, like:

<img width="256" height="132" alt="image" src="https://github.com/user-attachments/assets/56df6001-12b7-4880-b2ce-d8a3c5f8e3e7" />

It won't stop editors from trying to save multiple lines, like:

<img width="261" height="167" alt="image" src="https://github.com/user-attachments/assets/e817359d-791b-46e2-ad51-c5dfbfb75640" />

But upon save the HTML purifier configuration will strip the paragraph tags, so you end up with:

<img width="340" height="129" alt="image" src="https://github.com/user-attachments/assets/b269c34a-a205-4a26-88fc-58a19f90648b" />

It was the cleanest, safest approach I could find.

**Comment by mihob** [comment]
Yes, as far as I know, there is no other option without code adjustments. But I don't think that's a good solution.

The behaviour is confusing for the customer, and the unnecessary toolbar bloats the interface.
